zookeeper-server/src/main/java/org/apache/zookeeper/backup/NFSBackupStorage.java

```diff
@@ -0,0 +1,324 @@
+   * @param backupConfig The information and settings about backup storage, to be set as a part of ZooKeeper server config
+   */
+  public NFSBackupStorage(BackupConfig backupConfig) {
+    _backupConfig = backupConfig;
+    _fileRootPath =
+        String.join(PATH_SEPARATOR, _backupConfig.getMountPath(), _backupConfig.getNamespace());
+  }
+
+  @Override
+  public BackupFileInfo getBackupFileInfo(File file) throws IOException {
+    String backupFilePath = constructBackupFilePath(file.getName());
+    File backupFile = new File(backupFilePath);
+
+    if (!backupFile.exists()) {
+      throw new BackupException(
+          "Backup file " + file.getName() + " does not exist, could not get file info.");
+    }
+
+    BasicFileAttributes fileAttributes =
+        Files.readAttributes(Paths.get(backupFilePath), BasicFileAttributes.class);
+    return new BackupFileInfo(backupFile, fileAttributes.lastModifiedTime().toMillis(),
+        fileAttributes.size());
+  }
+
+  @Override
+  public List<BackupFileInfo> getBackupFileInfos(File path, String prefix) throws IOException {
+    String backupDirPath = constructBackupFilePath(path.getPath());
+    File backupDir = new File(backupDirPath);
+
+    if (!backupDir.exists()) {
+      throw new BackupException(
+          "Backup directory " + path.getPath() + " does not exist, could not get file info.");
+    }
+
+    // Get a list of files whose file names start with the provided prefix
+    FilenameFilter fileFilter = (dir, name) -> name.startsWith(prefix);
+    File[] files = backupDir.listFiles(fileFilter);
+    if (files == null) {
+      throw new BackupException("The provided directory path " + path.getPath()
+          + " is invalid, could not get file info.");
+    }
+
+    // Read the file info and add to the list. If an exception is thrown, the entire operation will fail
+    List<BackupFileInfo> backupFileInfos = new ArrayList<>();
+    for (File file : files) {
+      BasicFileAttributes fileAttributes =
+          Files.readAttributes(Paths.get(file.getPath()), BasicFileAttributes.class);
+      backupFileInfos.add(new BackupFileInfo(file, fileAttributes.lastModifiedTime().toMillis(),
+          fileAttributes.size()));
+    }
+    return backupFileInfos;
+  }
+
+  @Override
+  public List<File> getDirectories(File path) throws IOException {
+    String backupDirPath = constructBackupFilePath(path.getPath());
+    File backupDir = new File(backupDirPath);
+
+    if (!backupDir.exists()) {
+      throw new BackupException(
+          "Backup directory " + path.getPath() + " does not exist, could not get directory list.");
+    }
+
+    // Filter out all the files which are directories
+    FilenameFilter fileFilter = (dir, name) -> new File(dir, name).isDirectory();
+    File[] dirs = backupDir.listFiles(fileFilter);
+
+    if (dirs == null) {
+      throw new BackupException("The provided directory path " + path.getPath()
+          + " is invalid, could not get directory list.");
+    }
```

Comment:
Again, is there a reason we're throwing an exception over returning an empty list?

Reply:
Please refer to my last reply. Null means the path is invalid.

Reply:
When would the path be invalid?

Reply:
Please see my reply above.